policy: move PolicyMapState into policy package

ianvernon opened the pull request:
This is part of a broader effort to move most of the policy computation logic into the policy package itself, and out of endpoint. The idea is that in the future, we will have a single 'policy' object, which will be a field within `Endpoint`. The endpoint will consume this type, but will not be involved in its creation - that will be the responsibility of the policy package. This will make unit-testing policy generation easier, and will hide implementation details away from the endpoint package.

I don't like putting something related closely to the datapath implementation within the policy package, but in order to unify everything in one 'Policy' object for an endpoint, I think this is a necessary step to do so. Eventually, translation to `PolicyMapState` can be done via some translation into a datapath representation of the generated policy.

No functional changes are intended here.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6275)
<!-- Reviewable:end -->

Commit:
move PolicyMapState to policy package

This allows for it to be part of the Policy type within the policy package. This
is not an ideal change, since it exposes datapath representation into the policy
package, but this is an iterative step to unify policy computation within
Cilium. This will eventually be changed in the future.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: move determination of allow localhost / world into policy package

Now that PolicyMapState is within the policy package, refactor some functions
which perform operations directly on this type to be within said package.

Signed-off by: Ian Vernon <ian@cilium.io>

nebril reviewed (approved):
LGTM aside from hound comments

aanm reviewed (dismissed):
discussed offline

Commit:
policy: remove policymap dependency

Instead, duplicate the PolicyKey type in this package. The key difference
between this type and the PolicyKey type in pkg/maps/policymap, is that
the DestPort field is in host-byte order in the version in pkg/policy.
This avoids pulling in the BPF dependencies into pkg/policy, and also
continues the work to create an intermediate representation of policy
that is not directly tied to pkg/bpf. The duplicated type is renamed to
Key to avoid stuttering in the type name with the package name.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer reviewed (commented):
This seems fine, just one thing it'd be nice to reduce the stuttering as the hound bot is suggesting.

tgraf commented on pkg/policy/policymapstate.go:
```diff
@@ -0,0 +1,75 @@
+	}
+}
+
+// determineAllowLocalhost determines whether communication should be allowed to
```
This is legit

ianvernon replied:
Fixed.

tgraf reviewed (dismissed):
Looks awesome overall.

aanm commented on pkg/policy/mapstate.go:
```diff
@@ -0,0 +1,29 @@
+
+// MapStateEntry is the configuration associated with a PolicyKey in a
+// MapState. This is a minimized version of policymap.PolicyEntry.
+type MapStateEntry struct {
```
I think this structure needs a new name. This is storing proxy state AFAIK.

aanm commented on pkg/policy/mapstate.go:
```diff
@@ -27,3 +46,30 @@ type MapStateEntry struct {
+// This must be run after DetermineAllowLocalhost().
+//
+// For more information, see https://cilium.link/host-vs-world
+func (keys MapState) DetermineAllowFromWorld() {
```
When can we remove this? cc @joestringer

tgraf replied:
Let's do it in a dedicated PR and not mix it into this cleanup

aanm commented on pkg/policy/mapstate.go:
```diff
@@ -16,40 +16,55 @@ package policy
+type MapState map[Key]MapStateEntry
 
-// MapStateEntry is the configuration associated with a PolicyKey in a
+// Key is the userspace representation of a policy key in BPF. It is
```
Ok, we are getting somewhere. If this is supposed to be (B), then it should be in a different package. Here we are already mixing both A and C. There should be a package that can import both A and C so it can perform the translation from A to B and from B to C.

`[ (A) User Policy ] -> [ (B) Endpoint Policy ] -> [ (C) Datapath Policy ]`

The fields in this key structure are already datapath aware they have no relation with this policy package.

ianvernon replied:
@aanm this is meant to be an iterative change. I agree that we need to get to B, but I want to do this in a piecemeal, easy-to-review way. The reason I am doing this is so I can encapsulate what currently is translated policy for the endpoint in a single type, which I am doing in 
https://github.com/cilium/cilium/pull/6270 (which is based on top of this PR). In order to undo the current way we do things in policy translation, I am of the belief that if we can decouple the endpoint from policy translation (which is what this PR seeks to do), then we can get to B. After #6270 is reviewed / merged, I can then work to decouple what the generated policy for an endpoint is from the datapath implementation. But, to do that, I need to do some of the translation bits within the policy package, which for now, will include this `MapState` type.
